### ISSUE TITLE
GaussianSplatOptimizerMCMC progressively collapse

### ISSUE THREAD

**Issue by lorenzobagnol** [openvdb/fvdb-reality-capture#323]
## Summary

Training `GaussianSplatOptimizerMCMC` on a large-image-count scene (~400 images) produces reconstructions that progressively collapse during training — recognizable at early checkpoints, degenerating to unrecognizable "light beams" / near-empty space by the end — despite the exact same optimizer config working correctly on smaller scenes (10-50 images). Root cause appears to be an interaction between `opacity_regularization`'s fixed per-step pressure and `refine_every_epoch`'s epoch-based (i.e. dataset-size-dependent) scheduling.

## Environment

- `fvdb-reality-capture` version: 0.5.0
- Optimizer: `GaussianSplatOptimizerMCMCConfig`
- GPU: NVIDIA RTX A4500 (20GB)

## Mechanism

`regularization_loss()` (`gaussian_splat_optimizer_mcmc.py:358-375`) applies:

    loss = self._config.opacity_regularization * self._model.opacities.mean()

This gradient nudge is applied **every training step**, regardless of which image is being trained on (`batch_size=1`). Meanwhile, `refine()` (`gaussian_splat_optimizer_mcmc.py:323-356`) — the only mechanism that intervenes on Gaussians whose opacity has collapsed below `deletion_opacity_threshold`, via `_relocate()` — is scheduled in **epoch** units (`refine_every_epoch`), and one epoch = one full pass over the dataset = `N` steps at `batch_size=1`.

This means the number of **training steps between refine interventions** scales linearly with dataset size (`steps_per_refine = refine_every_epoch * N`), while opacity-regularization pressure accumulates every step regardless of `N`. As `N` grows, opacity regularization has proportionally more unchecked steps to push the population toward the deletion threshold before `refine()` can intervene — even though the *reconstruction* signal per Gaussian per refine-interval stays roughly constant (since `refine_every_epoch` epochs ≈ a fixed fraction of "full looks" at the scene, independent of `N`).

## Reproduction

**Run A (400 images, default `opacity_regularization=0.01`, native `refine_every_epoch=0.9`):**
Reconstruction is already degenerate ("light beams") by the 20% checkpoint, decaying to near-empty space by the end of training.

Disabling `opacity_regularization` (`=0`) or reducing it 10x (`=0.001`) fully resolved the collapse on this scene.

**Run B (50 images — a scene that has always worked fine at default settings — `opacity_regularization=0.01` unchanged, `refine_every_epoch` inflated to `7.2` to match Run A's per-interval step count of ~360 steps/refine, all other config identical):**

    [mcmc-refine-log] refine_count=1  num_gaussians_before=93,045    opacity_mean_before=0.76062
    [mcmc-refine-log] refine_count=5  num_gaussians_before=136,225   opacity_mean_before=0.34786
    [mcmc-refine-log] refine_count=10 num_gaussians_before=219,389   opacity_mean_before=0.24335
    [mcmc-refine-log] refine_count=15 num_gaussians_before=353,324   opacity_mean_before=0.19884
    [mcmc-refine-log] refine_count=20 num_gaussians_before=569,030   opacity_mean_before=0.14928
    [mcmc-refine-log] refine_count=26 num_gaussians_before=1,008,068 opacity_mean_before=0.10846

`opacity_mean` decays monotonically across the entire run (0.76 → 0.11), on a dataset that has never shown this behavior at native `refine_every_epoch=0.9`. Final reconstruction quality was degraded (though the object remained recognizable — likely because inflating `refine_every_epoch` also reduced the *total* number of refine events across the run, 26 here vs. ~208 for Run A's native schedule, since `max_epochs`/`refine_start_epoch`/`refine_stop_epoch` were not scaled proportionally — a fully matched reproduction would scale all four epoch-based parameters together and hasn't been tested yet).

This isolates the cause to **step-count-per-refine-interval**, not dataset size per se: dataset size only matters because it happens to be the multiplier on step-count in the library's epoch-based scheduling.

**Comment by lorenzobagnol** [comment]
## Update: opacity_regularization's magnitude matters independently of refine cadence

Follow-up after further testing. I framed the collapse as caused by `refine_every_epoch`'s epoch-based (dataset-size-dependent) scheduling diluting reconstruction correction relative to constant `opacity_regularization` pressure. That mechanism is still plausible, but new data shows it isn't the whole story — `opacity_regularization`'s absolute value matters on its own, even at unmodified cadence.

### Test 1: matching small-scene cadence made things worse

We shrank `refine_every_epoch` on the ~400-image scene (to ~0.08-0.11) so steps-per-refine matched the ~45 steps/refine that small (10-50 image) scenes get natively — refining roughly 8-9x more often. This produced a faster, more violent collapse than the original bug: by refine_count=77-79, population saturates `max_gaussians=4,000,000` with `num_relocated_dead` exceeding 3,000,000 per single refine step (>75% of the population dying/relocating at once), `opacity_mean` pinned near the deletion floor. (`opacity_regularization=0.002` here.)

### Test 2: lowering opacity_regularization further, at normal cadence, fixed it

Reverting `refine_every_epoch` to ~0.95 (essentially the library default — not scaled for dataset size) and lowering `opacity_regularization` to `0.0002` (50x below the `0.01` default) produced a run that stabilizes and recovers after saturating `max_gaussians`: `opacity_mean` climbs back from ~0.10 to ~0.15 and `num_relocated_dead` trends down (from ~540k to ~180k per step) over refine_count=78→105, at essentially the same steps-per-refine (~380) that caused collapse in the original report at `opacity_regularization=0.01`/`0.002`.

### Revised understanding

Both mechanisms likely compound, and we haven't cleanly separated them:

`opacity_regularization`'s absolute coefficient matters independently of cadence. Holding steps-per-refine roughly fixed (~360-380) on the same scene, we could swing the outcome from total collapse to stable recovery purely by lowering the coefficient 10-50x below values we'd previously tried. We suspect this is mediated by Gaussian population size — larger/denser scenes need MCMC to sustain much larger populations, and `_relocate()`/`_sample_add()`'s death/relocation dynamics may destabilize above some population scale unless `opacity_regularization` is well below the library default — but this isn't confirmed directly.

**Comment by harrism** [comment]
Thanks @lorenzobagnol, any chance you can provide a repro with a public dataset? That question applies to all the issues you have filed. Thanks!
